### PR TITLE
feat(optimizer): support like expression rewrite

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Support `like` expression rewrite.
-- `col like 'ABC'` => `col = 'ABC'`
-- `col like 'ABC%'` => `col >= 'ABC' and col < 'ABD'`
-- `col like 'ABC%E'` => `col >= 'ABC' and col < 'ABD' and col like 'ABC%E'`
- After rewriting, we can benefit from index selection.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>
#7855

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/index_selection.yaml [hunk 1]
```diff
@@ -608,3 +608,33 @@
...
+    BatchExchange { order: [], dist: Single }
+    └─BatchFilter { predicate: Like(idx.a, 'ABC_E':Varchar) }
+      └─BatchScan { table: idx, columns: [idx.a, idx.b, idx.c, idx.d], scan_ranges: [idx.a >= Utf8("ABC") AND idx.a <= Utf8("ABD")], distribution: UpstreamHashShard(idx.a) }
+- sql: |
+    create table t1 (a varchar, b int, c int, d int);
+    create index idx on t1(a);
+    select * from t1 where a like 'ABC'
+  batch_plan: |
+    BatchExchange { order: [], dist: Single }
+    └─BatchScan { table: idx, columns: [idx.a, idx.b, idx.c, idx.d], scan_ranges: [idx.a = Utf8("ABC")], distribution: UpstreamHashShard(idx.a) }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/index_selection.yaml:640]: Test cases.

**Comment by chenzl25** [comment]
The new project format would cause lots of existing plans changed after introducing a new rule for optimizer. :hot_face: cc @fuyufjh

File: src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+use super::{BoxedRule, Rule};
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, ExprVisitor, FunctionCall, Literal};
+use crate::optimizer::plan_node::{ExprRewritable, LogicalFilter};
+use crate::optimizer::PlanRef;
+
+pub struct RewriteLikeExprRule {}
+impl Rule for RewriteLikeExprRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let filter: &LogicalFilter = plan.as_logical_filter()?;
+        let mut has_like = HasLikeExprVisitor {};
```
Comment (xxchan) [inline src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs:31]: Why do we visit first? Is it for avoiding clone?

File: src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+use risingwave_common::types::ScalarImpl;
+use risingwave_connector::source::DataType;
+
+use super::{BoxedRule, Rule};
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, ExprVisitor, FunctionCall, Literal};
+use crate::optimizer::plan_node::{ExprRewritable, LogicalFilter};
+use crate::optimizer::PlanRef;
+
+pub struct RewriteLikeExprRule {}
```
Comment (xxchan) [inline src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs:27]: What about adding a rustdoc for it? 😄

File: src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                        ],
+                        DataType::Boolean,
+                    ).into()
+                }
+            } else {
+                FunctionCall::new_unchecked(func_type, inputs, ret).into()
+            }
+        } else {
+            FunctionCall::new_unchecked(func_type, inputs, ret).into()
+        }
```
Comment (xxchan) [inline src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs:152]: style: Using `let else` to put short brunches in the front might be slightly better. (and maybe return `func_call.into()`?)

File: src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    (Some(idx), None) => idx,
+                    (None, Some(idx)) => idx,
+                    (None, None) => {
+                        return FunctionCall::new_unchecked(ExprType::Equal, inputs, ret).into()
+                    }
+                };
+
+                if idx == 0 {
+                    return FunctionCall::new_unchecked(ExprType::Equal, inputs, ret).into()
+                }
```
Comment (xxchan) [inline src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs:96]: Should it be:
```suggestion
                    (None, None) => {
                        return FunctionCall::new_unchecked(func_type, inputs, ret).into()
                    }
                };

                if idx == 0 {
                    return FunctionCall::new_unchecked(func_type, inputs, ret).into()
                }
```
? (or `func_call.into()` + clone)

It seems wrong

```
dev=> explain select * from t where x like '_';
                    QUERY PLAN                    
--------------------------------------------------
 BatchExchange { order: [], dist: Single }
 └─BatchFilter { predicate: (t.x = '_':Varchar) }
   └─BatchScan { table: t, columns: [x] }
(3 rows)
```

**Review by xxchan — COMMENTED** [review]
others lgtm

File: src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs (hunk 2 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/rule/rewrite_like_expr_rule.rs:31]: Yes, if the filter doesn't contains any like expression, just ignore it.

**Comment by BugenZhao** [comment]
> The new project format would cause lots of existing plans changed after introducing a new rule for optimizer. 🥵 cc @fuyufjh

Is it possible to make the expression ID local and continuous? So that as long as the plan itself is not changed, we won't get diff on the explain result.

File: src/frontend/planner_test/tests/testdata/index_selection.yaml [hunk 6]
```diff
@@ -637,4 +637,5 @@
     select * from t1 where a like 'ABC'
   batch_plan: |
     BatchExchange { order: [], dist: Single }
-    └─BatchScan { table: idx, columns: [idx.a, idx.b, idx.c, idx.d], scan_ranges: [idx.a = Utf8("ABC")], distribution: UpstreamHashShard(idx.a) }
+    └─BatchFilter { predicate: Like(t1.a, 'ABC':Varchar) }
+      └─BatchScan { table: t1, columns: [t1.a, t1.b, t1.c, t1.d], distribution: SomeShard }
```
Comment (xxchan) [inline src/frontend/planner_test/tests/testdata/index_selection.yaml:641]: Why is this changed.. 🤔

**Review by xxchan — APPROVED** [review]
